### PR TITLE
Implement missing Influx aggregations

### PR DESCRIPTION
**Pull request by jkandasa**

### CODE REVIEW

**Comment by pilhuhn** [comment]
There are checkstyle errors

/home/travis/build/hawkular/hawkular-metrics/rest-servlet/src/main/java/org/rhq/metrics/restServlet/influx/InfluxSeriesHandler.java:372: Line has trailing spaces.
/home/travis/build/hawkular/hawkular-metrics/rest-servlet/src/main/java/org/rhq/metrics/restServlet/influx/InfluxSeriesHandler.java:443: Line is longer than 120 characters (found 207).
/home/travis/build/hawkular/hawkular-metrics/rest-servlet/src/main/java/org/rhq/metrics/restServlet/influx/InfluxSeriesHandler.java:452: Line is longer than 120 characters (found 210).

File: rest-servlet/src/main/java/org/hawkular/metrics/restServlet/influx/query/validation/AggregationFunction.java [hunk 1]
```diff
@@ -44,7 +44,17 @@
...
+    /** */
+    TOP("top"),
+    /** */
+    BOTTOM("bottom"),
+    /** */
+    HISTOGRAM("histogram"),
+    /** */
+    MODE("mode"),
+    /** */
+    STDDEV("stddev");
```
Comment (tsegismont) [inline rest-servlet/src/main/java/org/hawkular/metrics/restServlet/influx/query/validation/AggregationFunction.java]: All functions which takes extra arguments (top, bottom, histogram) should have a dedicated validator. See how PERCENTILE is declared.

File: rest-servlet/src/main/java/org/hawkular/metrics/restServlet/influx/InfluxSeriesHandler.java [hunk 2]
```diff
@@ -435,11 +437,65 @@ private String getColumnName(SelectQueryDefinitions queryDefinitions) {
                     NumberFunctionArgument argument = (NumberFunctionArgument) aggregationFunctionArguments.get(1);
                     retVal = quantil(list, argument.getDoubleValue());
                     break;
+                case TOP:
+                    isMultipleValues = true;
+                    if (!list.isEmpty() && aggregationFunctionArguments.size() > 0) {
+                        int numberOfTopElement = list.size() >
```
Comment (tsegismont) [inline rest-servlet/src/main/java/org/hawkular/metrics/restServlet/influx/InfluxSeriesHandler.java]: You shouldn't check the argument list here. If you attach a validator when declaring the function, the parsing tool will check the arguments for you.

File: rest-servlet/src/main/java/org/hawkular/metrics/restServlet/influx/InfluxSeriesHandler.java [hunk 3]
```diff
@@ -435,11 +437,65 @@ private String getColumnName(SelectQueryDefinitions queryDefinitions) {
...
+                            Integer.valueOf(aggregationFunctionArguments.get(0).toString()) ?
```
Comment (tsegismont) [inline rest-servlet/src/main/java/org/hawkular/metrics/restServlet/influx/InfluxSeriesHandler.java]: The first function argument is the metric name. What you want is the second argument. Besides, you don't need this conversion. Just cast the argument to the appropriate type (the validator should have checked the type earlier):
```java
NumberFunctionArgument argument = (NumberFunctionArgument) aggregationFunctionArguments.get(1);
```

See https://github.com/hawkular/hawkular-metrics/pull/148/files#diff-9ca02f4743984c633709aa4fbd223720L435

File: rest-servlet/src/main/java/org/hawkular/metrics/restServlet/influx/InfluxSeriesHandler.java [hunk 4]
```diff
@@ -435,11 +437,65 @@ private String getColumnName(SelectQueryDefinitions queryDefinitions) {
...
+                        int numberOfBottomElement = list.size() >
+                            Integer.valueOf(aggregationFunctionArguments.get(0).toString()) ?
+                            Integer.valueOf(aggregationFunctionArguments.get(0).toString()) : list.size();
+                        for(int elementPos = (numberOfBottomElement-1); elementPos>=0; elementPos--){
+                            retList.add(list.get(elementPos));
+                        }
+                    }
+                    break;
+                case HISTOGRAM:
+                case MODE:
```
Comment (tsegismont) [inline rest-servlet/src/main/java/org/hawkular/metrics/restServlet/influx/InfluxSeriesHandler.java]: I'm not sure how the  `histogram` and `mode` functions behave. The documentation is quite vague:
http://influxdb.com/docs/v0.8/api/aggregate_functions.html

Have you tried to find the answer on the online InfluxDB service?

I'd be fine with leaving those two functions out of the scope for now.

**Comment by tsegismont** [comment]
@jkandasa Thank you Jeeva for the PR. Other than the particular issues I've pointed inline, could you add some tests in the rest-tests module for the new functions?

**Comment by jkandasa** [comment]
@tsegismont Thank you for the review and comments. I will fix it soon.

File: rest-servlet/src/main/java/org/hawkular/metrics/restServlet/influx/InfluxSeriesHandler.java (hunk 4 above)
Comment (jkandasa) [inline rest-servlet/src/main/java/org/hawkular/metrics/restServlet/influx/InfluxSeriesHandler.java]: `histogram` and `mode`, not changed anything.

**Comment by jkandasa** [comment]
@tsegismont Hi Thomas, Kindly review my changes. Thank you.

**Comment by tsegismont** [comment]
Looks good, thank you very much Jeeva!

**Comment by tsegismont** [comment]
Fixes issue #148

**Comment by pilhuhn** [comment]
Thank you Jeeva

**Pull request merged**